blog/2020-11-10-streaming-vitess-at-bolt.adoc

```diff
@@ -0,0 +1,181 @@
+* Connection timeout and transaction timeout can be configured.
+* It has minimal downtime when doing link:https://vitess.io/docs/user-guides/configuration-advanced/resharding/[resharding] operations.
+* Its VStream feature can be used by downstream CDC applications to read change events from Vitess.
+
+== Streaming Vitess Options
+
+The ability to capture data changes and publish them to Apache Kafka was one of the requirements for adopting Vitess at Bolt. There were several different options we’ve considered.
+
+=== Option 1: Using Debezium MySQL Connector
+
+Applications connect to Vitess VTGate to send queries. VTGate supports MySQL protocol and has a SQL parser. You can use any MySQL client (e.g. JDBC) to connect to VTGate, which redirects your query to the correct shard and returns the result to your client.
+
+However, VTGate is not equal to a MySQL instance, it is rather a stateless proxy to various MySQL instances. For the MySQL connector to receive change events, the Debezium MySQL connector needs to connect to a real MySQL instance. To make it more obvious, VTGate also has some known link:https://vitess.io/docs/reference/compatibility/mysql-compatibility/[compatibility] issues, which makes connecting to VTGate different from MySQL.
+
+Another option is to use the Debezium MySQL Connector to connect directly to the underlying MySQL instances of different shards. It has its advantages and disadvantages.
+
+One advantage is that for an unsharded keyspace (Vitess's terminology for a database), the MySQL Connector can continue to work correctly and we don't need to include additional logic or specific implementation. It should just work fine.
+
+One of the biggest disadvantages is that resharding operations would become more complex. For example, the GTID of the original MySQL instance would change when resharded, and the MySQL connector depends on the GTID to work correctly. We also believe that having the MySQL connector connected directly to each underlying MySQL instance defies the purpose of Vitess's operational simplicity as a new connector has to be added (or removed) each time resharding is done. Not to mention that such operation would lead to data duplication inside Kafka brokers.
+
+=== Option 2: Using JDBC Source Connector
+
+We've also considered using link:https://docs.confluent.io/current/connect/kafka-connect-jdbc/source-connector/index.html[JDBC Source Connector]. It allows sourcing data from any relational databases that support the JDBC driver into Kafka. Therefore, it is compatible with Vitess VTGate. It has its advantages and disadvantages as well.
+
+Advantages:
+
+* It is compatible with VTGate.
+* It handles Vitess resharding operation better. During resharding operation, reads are simply automatically redirected (by VTGate) to the target shards. It won't generate any duplicates or lose any data.
+
+Disadvantages:
```

Comment:
You'd also miss things like TX ids, potentially multiple updates in close timely proximity etc. Perhaps link to this blog post here which compares log-based vs. polling based CDC: https://debezium.io/blog/2018/07/19/advantages-of-log-based-change-data-capture/.

Reply:
That's a good point actually. Unfortunately, Vitess 8 (the latest stable version)'s change events only have VGTID, no TX ids yet. Does MySQL connector have TX ids?

Reply:
Thanks for the reference. I'll read the log-based vs. polling-based CDC blog post